source/common/protobuf/utility.cc

```diff
@@ -27,6 +27,37 @@ using namespace std::chrono_literals;
 namespace Envoy {
 namespace {
 
+absl::Status validateDurationNoThrow(const ProtobufWkt::Duration& duration) {
+  // Apply a strict max boundary to the `seconds` value to avoid overflow.
+  // Note that protobuf internally converts to nanoseconds.
+  // The kMaxSecondsValue = 9223372036, which is about 300 years.
+  constexpr int64_t kMaxSecondsValue = std::numeric_limits<int64_t>::max() / (1000 * 1000 * 1000);
+
+  if (duration.seconds() < 0 || duration.nanos() < 0) {
+    return absl::OutOfRangeError(
+        fmt::format("Expected positive duration: {}", duration.DebugString()));
+  }
+  if (!Protobuf::util::TimeUtil::IsDurationValid(duration)) {
+    return absl::OutOfRangeError(
+        fmt::format("Duration out-of-range according to Protobuf: {}", duration.DebugString()));
+  }
+  if (duration.nanos() > 999999999 || duration.seconds() > kMaxSecondsValue) {
```

Comment:
nit: it can still overflow with nano = 999,999,999 and secs = kMaxSecondsValue :)

Reply:
Good catch.
The max value previously used had that problem.
I've updated the maximal value to take this into account.

I've also modified the runtime guard to allow reverting to the "old" behavior (which wasn't good, but now at least there's a knob the reverts everything to the way it was).